consensus/cryptarchia-sync/core/src/sync.rs

```diff
@@ -0,0 +1,214 @@
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    error::Error,
+};
+
+use cryptarchia_engine::Slot;
+use futures::{Stream, StreamExt};
+use itertools::Itertools;
+use nomos_core::{block::AbstractBlock, header::HeaderId};
+use tracing::{debug, info};
+
+use crate::{fetcher, CryptarchiaSyncError, CryptarchiaSyncExt};
+
+#[async_trait::async_trait]
+pub trait CryptarchiaSyncExtInternal<BlockFetcher>: CryptarchiaSyncExt
+where
+    Self::Block: Send,
+    BlockFetcher: fetcher::BlockFetcher<Block = Self::Block> + Sync,
+{
+    /// Start a sync process.
+    async fn run(
+        mut self,
+        block_fetcher: &BlockFetcher,
+    ) -> Result<Self, Box<dyn Error + Send + Sync>> {
+        // Start the sync process from the current local tip slot.
+        let mut start_slot = self.tip_slot() + 1;
+        info!("Starting sync process from slot {start_slot:?}");
+
+        let mut rejected_blocks = HashSet::new();
+        // Run a forward sync, and trigger backfillings if orphan blocks are found.
+        // Repeat the forward sync if any backfilling was performed,
+        // in order to catch up with the updated peers.
+        loop {
+            let ForwardSyncOutput {
+                max_fetched_slot,
+                orphan_blocks,
+            } = self
+                .sync_forward(start_slot, block_fetcher, &mut rejected_blocks)
```

Comment:
what would happen if the first block has a missing parent and even though you fetch all the block (i.e. max_start_slot is >> start_slot) you didn't successfully apply any of those?